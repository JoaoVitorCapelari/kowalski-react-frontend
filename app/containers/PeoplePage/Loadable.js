/**
 *
 * Asynchronously loads the component for PeoplePage
 *
 */

import Loadable from 'react-loadable';

export default Loadable({
  loader: () => import('./index'),
  loading: () => null,
});
