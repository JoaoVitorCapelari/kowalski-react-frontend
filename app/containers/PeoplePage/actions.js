/*
 *
 * PeoplePage actions
 *
 */

import {
  OPEN_NEW_PERSON_FORM,
  CLOSE_NEW_PERSON_FORM,
  SUBMIT_NEW_PERSON_FORM,
  NEW_PERSON_SAVED,
  SUBMIT_NEW_PERSON_FORM_AND_CLOSE_IT,
  PERSON_SELECTED,
  START_PEOPLE_LOADING,
  END_PEOPLE_LOADING,
  START_FETCHING_NEW_PERSON_FORM_OPTIONS,
  END_FETCHING_NEW_PERSON_FORM_OPTIONS,
} from './constants';

export function openNewPersonForm() {
  return {
    type: OPEN_NEW_PERSON_FORM,
  };
}

export function closeNewPersonForm() {
  return {
    type: CLOSE_NEW_PERSON_FORM,
  };
}

export function submitNewPersonForm(formData) {
  return {
    type: SUBMIT_NEW_PERSON_FORM,
    payload: formData,
  };
}

export function submitNewPersonFormAndCloseIt(formData) {
  return {
    type: SUBMIT_NEW_PERSON_FORM_AND_CLOSE_IT,
    payload: formData,
  };
}

export function newPersonSaved(newPersonData) {
  return {
    type: NEW_PERSON_SAVED,
    payload: newPersonData,
  };
}

export function personSelected(projectCode) {
  return {
    type: PERSON_SELECTED,
    payload: projectCode,
  };
}

export function startPeopleLoading() {
  return {
    type: START_PEOPLE_LOADING,
  };
}

export function endPeopleLoading(payload) {
  return {
    type: END_PEOPLE_LOADING,
    payload,
  };
}

export function startFetchingNewPersonFormOptions() {
  return {
    type: START_FETCHING_NEW_PERSON_FORM_OPTIONS,
  };
}

export function endFetchingNewPersonFormOptions(payload) {
  return {
    type: END_FETCHING_NEW_PERSON_FORM_OPTIONS,
    payload,
  };
}
