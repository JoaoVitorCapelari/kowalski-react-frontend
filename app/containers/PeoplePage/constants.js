/*
 *
 * PeoplePage constants
 *
 */

export const DEFAULT_ACTION = 'app/PeoplePage/DEFAULT_ACTION';
export const NEW_PERSON_FORM_ID = 'app/PersonPage/NEW_PERSON_FORM_ID';

export const OPEN_NEW_PERSON_FORM = 'app/PeoplePage/OPEN_NEW_PERSON_FORM';
export const CLOSE_NEW_PERSON_FORM = 'app/PeoplePage/CLOSE_NEW_PERSON_FORM';
export const SUBMIT_NEW_PERSON_FORM = 'app/PeoplePage/SUBMIT_NEW_PERSON_FORM';
export const NEW_PERSON_SAVED = 'app/PeoplePage/NEW_PERSON_SAVED';
export const SUBMIT_NEW_PERSON_FORM_AND_CLOSE_IT = 'app/PeoplePage/SUBMIT_NEW_PERSON_FORM_AND_CLOSE_IT';
export const PERSON_SELECTED = 'app/PeoplePage/PERSON_SELECTED';
export const START_PEOPLE_LOADING = 'app/PeoplePage/START_PEOPLE_LOADING';
export const END_PEOPLE_LOADING = 'app/PeoplePage/END_PEOPLE_LOADING';
export const START_FETCHING_NEW_PERSON_FORM_OPTIONS = 'app/PeoplePage/START_FETCHING_NEW_PERSON_FORM_OPTIONS';
export const END_FETCHING_NEW_PERSON_FORM_OPTIONS = 'app/PeoplePage/END_FETCHING_NEW_PERSON_FORM_OPTIONS';
