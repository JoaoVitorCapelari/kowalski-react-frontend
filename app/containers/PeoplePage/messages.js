/*
 * PeoplePage Messages
 *
 * This contains all the text for the PeoplePage component.
 */
import { defineMessages } from 'react-intl';

export default defineMessages({
  title: {
    id: 'app.containers.PeoplePage.title',
    defaultMessage: 'People',
  },
  description: {
    id: 'app.containers.PeoplePage.description',
    defaultMessage: 'Manage or View System People',
  },
  newPersonFormTitle: {
    id: 'app.containers.ProjectPage.newPersonFormTitle',
    defaultMessage: 'Add person',
  },
  newPersonFormChooseAFile: {
    id: 'app.containers.ProjectPage.newPersonFormChooseAFile',
    defaultMessage: 'Choose a file',
  },
  newPersonFormFileHelpText: {
    id: 'app.containers.ProjectPage.newPersonFormFileHelpText',
    defaultMessage: 'Your file must be in .jpeg or .png format',
  },
  newPersonFormLabelUsername: {
    id: 'app.containers.ProjectPage.newPersonFormLabelUsername',
    defaultMessage: 'Username',
  },
  newPersonFormLabelRole: {
    id: 'app.containers.ProjectPage.newPersonFormLabelRole',
    defaultMessage: 'Role',
  },
  newPersonFormLabelName: {
    id: 'app.containers.ProjectPage.newPersonFormLabelName',
    defaultMessage: 'Name',
  },
  newPersonFormLabelEmail: {
    id: 'app.containers.ProjectPage.newPersonFormLabelEmail',
    defaultMessage: 'Email',
  },
  newPersonFormButtonSaveAndAddNew: {
    id: 'app.containers.ProjectPage.newPersonFormButtonSaveAndAddNew',
    defaultMessage: 'Save and add new',
  },
  newPersonFormButtonAdd: {
    id: 'app.containers.ProjectPage.newPersonFormButtonAdd',
    defaultMessage: 'Add',
  },
  newPersonFormButtonCancel: {
    id: 'app.containers.ProjectPage.newPersonFormButtonCancel',
    defaultMessage: 'Cancel',
  },
  roleManager: {
    id: 'app.containers.ProjectPage.roleManager',
    defaultMessage: 'Manager',
  },
  roleUser: {
    id: 'app.containers.ProjectPage.roleUser',
    defaultMessage: 'User',
  },
  roleAuditor: {
    id: 'app.containers.ProjectPage.roleAuditor',
    defaultMessage: 'Auditor',
  },
  roleAdministrator: {
    id: 'app.containers.ProjectPage.roleAdministrator',
    defaultMessage: 'Administrator',
  },
  newPersonFormSelectARole: {
    id: 'app.containers.ProjectPage.newPersonFormSelectARole',
    defaultMessage: 'Select a Role',
  },
  buttonAddPerson: {
    id: 'app.containers.ProjectPage.buttonAddPerson',
    defaultMessage: 'Add person',
  },
  resourceNamePlural: {
    id: 'app.containers.PeoplePage.resourceNamePlural',
    defaultMessage: 'people',
  },
});
