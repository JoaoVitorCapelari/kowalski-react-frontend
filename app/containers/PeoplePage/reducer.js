/*
 *
 * PeoplePage reducer
 *
 */

import { fromJS } from 'immutable';
import {
  OPEN_NEW_PERSON_FORM,
  CLOSE_NEW_PERSON_FORM,
  START_PEOPLE_LOADING,
  END_PEOPLE_LOADING,
} from './constants';

const initialState = fromJS({
  people: [],
  personFormOpen: false,
  isLoadingPeople: false,
  loadingPeopleErrorMsg: '',
});

function peoplePageReducer(state = initialState, { type, payload }) {
  switch (type) {
    case OPEN_NEW_PERSON_FORM:
      return state.set('personFormOpen', true);
    case CLOSE_NEW_PERSON_FORM:
      return state.set('personFormOpen', false);
    case START_PEOPLE_LOADING:
      return state.set('isLoadingPeople', true);
    case END_PEOPLE_LOADING:
      return state
        .set('isLoadingPeople', false)
        .set('loadingPeopleErrorMsg', payload.success ? '' : payload.errorMsg)
        .set('people', payload.success ? payload.data : state.get('people'));
    default:
      return state;
  }
}

export default peoplePageReducer;
