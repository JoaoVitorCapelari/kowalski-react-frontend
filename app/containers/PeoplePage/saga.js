import { takeEvery, call, put } from 'redux-saga/effects';
import { fromJS } from 'immutable';

import {
  reset,
  stopSubmit,
  startSubmit,
} from 'redux-form';

import {
  SUBMIT_NEW_PERSON_FORM,
  SUBMIT_NEW_PERSON_FORM_AND_CLOSE_IT,
  NEW_PERSON_FORM_ID,
  CLOSE_NEW_PERSON_FORM,
  PERSON_SELECTED,
  START_PEOPLE_LOADING,
} from './constants';

import {
  startPeopleLoading,
  closeNewPersonForm,
  endPeopleLoading,
} from './actions';

import {
  requestErrorReceived,
} from '../App/actions';

import { getPeople, registerPerson } from '../../support/backend/KowalskiBackendClient';
import { genCommonReqConfig } from '../../support/backend/utils';

export function* handleSubmit({ payload }) {
  const successful = yield submitPersonForm({ payload });
  if (successful) {
    yield clearPersonForm();
  }
}

export function* handleSubmitAndCloseForm({ payload }) {
  const successful = yield submitPersonForm({ payload });
  if (successful) {
    yield put(closeNewPersonForm());
  }
}

export function* handlePersonSelected() {
  // Do nothing for now -> should open the person dialog allowing for edition in the future
}

export function* submitPersonForm({ payload }) {
  yield put(startSubmit(NEW_PERSON_FORM_ID));
  try {
    yield call(registerPerson, {
      ...genCommonReqConfig(),
      personData: payload.toJSON(),
    });
    yield put(stopSubmit(NEW_PERSON_FORM_ID));
    yield put(startPeopleLoading());
    return true;
  } catch (e) {
    yield put(requestErrorReceived({
      error: e,
      dispatchOnAuthError: [], // Redux-form will automatically clean up state in this case
      dispatchOnOtherErrors: [
        stopSubmit(
          NEW_PERSON_FORM_ID,
          { _error: 'There was an error while trying to communicate with the server =(' }
        ),
      ],
    }));
  }

  return false;
}

export function* clearPersonForm() {
  yield put(reset(NEW_PERSON_FORM_ID));
}

export function* handlePeopleLoading() {
  try {
    const people = yield call(
      getPeople,
      genCommonReqConfig(),
    );
    yield put(endPeopleLoading({ success: true, data: fromJS(people) }));
  } catch (e) {
    yield put(requestErrorReceived({
      error: e,
      dispatchOnAuthError: [
        endPeopleLoading({ success: false, errorMsg: '' }),
      ],
      dispatchOnOtherErrors: [
        endPeopleLoading({ success: false, errorMsg: 'There was an error while trying to communicate with the server =(' }),
      ],
    }));
  }
}

// Individual exports for testing
export default function* defaultSaga() {
  yield takeEvery(SUBMIT_NEW_PERSON_FORM, handleSubmit);
  yield takeEvery(SUBMIT_NEW_PERSON_FORM_AND_CLOSE_IT, handleSubmitAndCloseForm);
  yield takeEvery(CLOSE_NEW_PERSON_FORM, clearPersonForm);
  yield takeEvery(PERSON_SELECTED, handlePersonSelected);
  yield takeEvery(START_PEOPLE_LOADING, handlePeopleLoading);
}
