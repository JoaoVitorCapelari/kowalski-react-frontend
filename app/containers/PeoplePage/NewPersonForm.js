import React from 'react';
import styled from 'styled-components';
import PropType from 'prop-types';
import { Field, reduxForm } from 'redux-form/immutable';

import InputField from 'components/InputField/Loadable';
import TextAreaField from 'components/TextAreaField/Loadable';

import { required } from '../../support/forms/validation';
import { NEW_PERSON_FORM_ID } from './constants';
import { formatDate } from '../../support/backend/formatters';

const Wrapper = styled.div`
  display: flex;
  flex-grow: 1;
  flex-direction: column;
  justify-content: center;
  align-content: center;
`;

const FormTitle = styled.div`
`;

const FormActions = styled.div`
  display: flex;
  justify-content: center;
  margin: 2rem;
`;

const H1 = styled.h1`
  text-align: center;
`;

const FormAction = styled.button`
  font-size: 1.1rem;
  width: 100%;
`;

const FormActionWrapper = styled.div`
  width: 100%;
`;

function NewPersonForm(props) {
  const { error, isSubmitting, onCancel, handleSubmit, onAdd, onSaveAndAddNew } = props;

  const myOnSubmit = (formData, func) => (
    func(formData
      .set('startDate', formatDate(new Date()))
      .set('endDate', formatDate(new Date()))
      .set('code', formData.get('name'))
    )
  );

  return (
    <Wrapper className="box">
      <form onSubmit={handleSubmit((formData) => myOnSubmit(formData, onAdd))}>
        <FormTitle><H1>Create new project</H1></FormTitle>

        <Field
          name="name"
          id="name"
          component={InputField}
          label="Project Name"
          validate={[required]}
        />

        <Field
          name="description"
          id="description"
          component={TextAreaField}
          label="Project Description"
          validate={[required]}
        />

        {
          error &&
          <div className="control" style={{ marginTop: '1rem' }}>
            <article className="message is-danger">
              <div className="message-body">
                { error }
              </div>
            </article>
          </div>
        }

        <FormActions>
          <FormActionWrapper className="control">
            <FormAction
              className="button"
              type="button"
              disabled={isSubmitting}
              onClick={handleSubmit((formData) => myOnSubmit(formData, onSaveAndAddNew))}
            >
              Save and add new
            </FormAction>
          </FormActionWrapper>
          <FormActionWrapper className="control">
            <FormAction type="submit" className={`button is-primary ${isSubmitting ? 'is-loading' : ''}`} disabled={isSubmitting}>
              Add
            </FormAction>
          </FormActionWrapper>
          <FormActionWrapper className="control">
            <FormAction className="button" type="button" onClick={onCancel}>Cancel</FormAction>
          </FormActionWrapper>
        </FormActions>
      </form>
    </Wrapper>
  );
}

NewPersonForm.propTypes = {
  error: PropType.any,
  isSubmitting: PropType.any,
  onCancel: PropType.func,
  handleSubmit: PropType.func,
  onAdd: PropType.func,
  onSaveAndAddNew: PropType.func,
};

export default reduxForm({
  form: NEW_PERSON_FORM_ID,
})(NewPersonForm);
