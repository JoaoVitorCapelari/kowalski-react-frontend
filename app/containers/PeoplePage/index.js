/**
 *
 * PeoplePage
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import Modal from 'components/Modal/Loadable';

import injectSaga from 'utils/injectSaga';
import injectReducer from 'utils/injectReducer';
import makeSelectPeoplePage from './selectors';
import reducer from './reducer';
import saga from './saga';
import * as actions from './actions';
import NewPersonForm from './NewPersonForm';

const MainContainerWrapper = styled.div`
  display: flex;
  flex-direction: column;
`;

const margins = `
  margin-top: 2rem;
  margin-left: 4rem;
  margin-right: 4rem;
`;

const TitleBar = styled.div`
  ${margins}
  display: flex;
  flex-basis: 100%;
  align-items: center;
  justify-content: space-between;
`;

const PageTitle = styled.h1`
  display: block;
  font-weight:300px;
`;

const AddPersonButton = styled.button`
  width:226px;
  height:50px !important;
  border-radius:2px;
  border:1px solid #654EA3;
  color: #654EA3;
`;

const PersonListWrapper = styled.div`
  ${margins}
  display: flex;
  flex-grow: 1;
  flex-direction: column;
`;

export class PeoplePage extends React.Component { // eslint-disable-line react/prefer-stateless-function

  render() {
    const {
      openNewPersonForm,
      closeNewPersonForm,
      isNewPersonFormOpen,
    } = this.props;

    return (
      <MainContainerWrapper className="kowalski-react-basic-container">
        <Helmet>
          <title>PeoplePage</title>
          <meta name="description" content="Description of PeoplePage" />
        </Helmet>
        <TitleBar>
          <PageTitle>People</PageTitle>
          <AddPersonButton className="button" onClick={openNewPersonForm}>Add person</AddPersonButton>
        </TitleBar>
        <PersonListWrapper>
          { this.renderProjectsPanel() }
        </PersonListWrapper>
        <Modal active={isNewPersonFormOpen} onDismiss={closeNewPersonForm}>
          <NewPersonForm
            onAdd={this.props.submitNewPersonFormAndCloseIt}
            onSaveAndAddNew={this.props.submitNewPersonForm}
            onCancel={closeNewPersonForm}
          />
        </Modal>
      </MainContainerWrapper>
    );
  }
}

PeoplePage.propTypes = {
  openNewPersonForm: PropTypes.func,
  closeNewPersonForm: PropTypes.func,
  isNewPersonFormOpen: PropTypes.bool,
  submitNewPersonFormAndCloseIt: PropTypes.func,
  submitNewPersonForm: PropTypes.func,
};

const mapStateToProps = createStructuredSelector({
  peoplepage: makeSelectPeoplePage(),
});

const withConnect = connect(mapStateToProps, actions);
const withReducer = injectReducer({ key: 'peoplePage', reducer });
const withSaga = injectSaga({ key: 'peoplePage', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(PeoplePage);
