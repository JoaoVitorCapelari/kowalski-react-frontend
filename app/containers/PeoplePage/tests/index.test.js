import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { fromJS } from 'immutable';
import { IntlProvider } from 'react-intl';
import createHistory from 'history/createMemoryHistory';

import PeoplePage from '../index';
import { mockLocalStorage, unmockLocalStorage } from '../../../support/tests/stub';
import { translationMessages } from '../../../i18n';
import configureStore from '../../../configureStore';

const initialState = {};
const history = createHistory();
const store = configureStore(initialState, history);

describe('<PeoplePage />', () => {
  beforeEach(mockLocalStorage);
  afterEach(unmockLocalStorage);

  it('renders', () => {
    const renderedComponent = mount(
      <Provider store={store}>
        <IntlProvider locale="en" messages={translationMessages.en}>
          <PeoplePage
            openNewPersonForm={jest.fn()}
            closeNewPersonForm={jest.fn()}
            isNewPersonFormOpen={false}
            submitNewPersonFormAndCloseIt={jest.fn()}
            submitNewPersonForm={jest.fn()}
            startPeopleLoading={jest.fn()}
            people={fromJS([])}
            isLoadingPeople={false}
            personSelected={jest.fn()}
            loadingPeopleErrorMsg=""
            intl={{ formatMessage: (v) => v.toString() }}
          />
        </IntlProvider>
      </Provider>
    );
    expect(renderedComponent.html()).toMatchSnapshot();
  });
});
