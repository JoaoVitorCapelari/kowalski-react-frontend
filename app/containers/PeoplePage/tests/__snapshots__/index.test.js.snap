// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PeoplePage /> renders 1`] = `"<div class=\\"kowalski-react-basic-container PeoplePage__MainContainerWrapper-kgxPKm blPnBe\\"><!-- react-empty: 2 --><div class=\\"PeoplePage__TitleBar-dhiXIa gAyozb\\"><h1 class=\\"PeoplePage__PageTitle-dbaXBP iiJKgw\\"><span>People</span></h1></div><div class=\\"PeoplePage__PersonListWrapper-dEbFOF feLAOG\\"><div class=\\"css-jxiqlq\\"><div class=\\"css-1lt1sep\\"></div><div class=\\"css-1vahj0q\\"></div></div></div><!-- react-empty: 9 --></div>"`;
